src/core/credentials/transport/tls/grpc_tls_certificate_provider.h

```diff
@@ -208,6 +210,62 @@ class FileWatcherCertificateProvider final
   std::map<std::string, WatcherInfo> watcher_info_ ABSL_GUARDED_BY(mu_);
 };
 
+// Implements a provider that uses in-memory data that can be modified in a
+// thread-safe manner.
+class InMemoryCertificateProvider final : public grpc_tls_certificate_provider {
+ public:
+  static std::shared_ptr<InMemoryCertificateProvider> CreateCertificateProvider(
+      std::string root_certificates, PemKeyCertPairList pem_key_cert_pairs);
+  InMemoryCertificateProvider(std::string root_certificates,
+                              PemKeyCertPairList pem_key_cert_pairs);
+  InMemoryCertificateProvider(const InMemoryCertificateProvider&) = delete;
+  InMemoryCertificateProvider(InMemoryCertificateProvider&&) = delete;
+  InMemoryCertificateProvider& operator=(const InMemoryCertificateProvider&) =
+      delete;
+  InMemoryCertificateProvider& operator=(InMemoryCertificateProvider&&) =
+      delete;
+
+  RefCountedPtr<grpc_tls_certificate_distributor> distributor() const override {
+    return distributor_;
+  }
+
+  UniqueTypeName type() const override;
+
+  absl::Status ValidateCredentials() const;
+
+  int64_t TestOnlyGetRefreshIntervalSecond() const;
+
+  void UpdateRoot(std::string root_certificates);
+  void UpdateIdentity(const PemKeyCertPairList& pem_key_cert_pairs);
+
+ private:
+  struct WatcherInfo {
+    bool root_being_watched = false;
+    bool identity_being_watched = false;
+  };
+
+  int CompareImpl(const grpc_tls_certificate_provider* other) const override {
+    return QsortCompare(static_cast<const grpc_tls_certificate_provider*>(this),
+                        other);
+  }
+  RefCountedPtr<grpc_tls_certificate_distributor> distributor_;
+
+  std::shared_ptr<RootCertInfo> root_cert_info_;
+
+  // Guards pem_key_cert_pairs_, root_certificates_ and watcher_info_
+  // respectively.
+  mutable Mutex pem_cert_pairs_mu_;
```

Comment:
Please just use one mutex for all data members.  Using multiple mutexes opens the possibility of deadlocks by acquiring mutexes in different orders in different threads, and it doesn't really yield any significant performance wins in this case.